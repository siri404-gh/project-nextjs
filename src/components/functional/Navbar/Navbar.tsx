import React, { useState, FunctionComponent } from 'react';
import {
  AppBar,
  Typography,
  IconButton,
  makeStyles,
  Toolbar,
  Hidden,
  Avatar,
  Menu,
  MenuItem,
} from '@material-ui/core';
import Link from 'next/link';
import { AccountCircle, Menu as MenuIcon } from '@material-ui/icons';
import cn from 'classnames';

const useStyles = makeStyles(theme => ({
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
  },
  leftBtn: {
    padding: 6,
  },
  titleLeft: {
    flexGrow: 1,
  },
  titlebar: {
    display: 'flex',
    flexGrow: 1,
    cursor: 'pointer',
    justifyContent: 'left',
    alignItems: 'center',
    padding: 4,
  },
  title: {
    display: 'inline',
  },
  description: {
    display: 'inline',
    fontStyle: 'italic',
    marginRight: 6,
  },
  toolbar: {
    minHeight: 56,
    paddingLeft: 10,
    paddingRight: 10,
  },
  small: {
    width: theme.spacing(3),
    height: theme.spacing(3),
  },
}));

export interface Props {
  title?: string,
  description?: string,
  onTitlebarClick?: () => void,
  onLeftButtonClick?: () => void,
  isUserLoggedIn?: boolean,
  isUserPremium?: boolean,
  loginRedirectUrl?: string,
  avatarUrl?: string,
}

const Navbar: FunctionComponent<Props> = ({ title, description, onTitlebarClick, onLeftButtonClick, isUserLoggedIn, isUserPremium, loginRedirectUrl, avatarUrl }) => {
  const classes = useStyles();
  const [ anchorEl, setAnchorEl ] = useState(null);

  const onAvatarToggle = e => {
    e.stopPropagation()
    setAnchorEl(!anchorEl ? e.currentTarget : null);
  };

  return (
    <AppBar className={classes.appBar} elevation={0}>
      <Toolbar className={classes.toolbar}>
        <IconButton className={cn('Navbar__leftbutton', classes.leftBtn)} onClick={onLeftButtonClick} aria-hidden>
          <MenuIcon color='secondary' />
        </IconButton>
        <div className={cn('Navbar__titlebar', classes.titlebar)} onClick={onTitlebarClick}>
          <div className={classes.titleLeft}>
            <Hidden smUp>
              <Typography className={classes.title} noWrap variant='h6'>{description}</Typography>
            </Hidden>
            <Hidden xsDown>
              <Typography className={classes.title} noWrap variant='h6'>{title}</Typography>
             &nbsp; | &nbsp;
              <Typography className={classes.description} noWrap variant='caption'>{description}</Typography>
            </Hidden>
          </div>
          <IconButton aria-controls='simple-menu' onClick={onAvatarToggle} aria-hidden>
            {isUserLoggedIn
              ? <Avatar alt='avatar' src={avatarUrl} className={classes.small} />
              : <AccountCircle color={'secondary'} className={classes.small} />}
            <Menu
              id='simple-menu'
              anchorEl={anchorEl}
              keepMounted
              open={Boolean(anchorEl)}
              onClose={onAvatarToggle}>
              {isUserLoggedIn
                ? <Link href={'/api/auth/logout'}>
                  <MenuItem>Logout</MenuItem>
                </Link>
                : <Link href={`/api/auth/login?redirectTo=${loginRedirectUrl}`}>
                  <MenuItem>Login</MenuItem>
                </Link>}
            </Menu>
          </IconButton>
        </div>
      </Toolbar>
    </AppBar>
  );
};

Navbar.defaultProps = {
  title: '',
  description: '',
  onLeftButtonClick: /* istanbul ignore next */ () => null,
  onTitlebarClick: /* istanbul ignore next */ () => null,
  loginRedirectUrl: '/',
};

export default Navbar;
