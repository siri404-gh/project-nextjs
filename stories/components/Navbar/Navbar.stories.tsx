import React from 'react';
import { Story } from '@storybook/react/types-6-0';

import Navbar, { Props } from '@/components/functional/Navbar/Navbar';

export default {
  title: 'Navbar',
  component: Navbar,
};

const Template: Story<Props> = args => <Navbar {...args} />;

export const FirstStory = Template.bind({});

FirstStory.args = {};
