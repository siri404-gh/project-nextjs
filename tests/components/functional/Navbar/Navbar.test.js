import { shallow } from 'enzyme';

import Navbar from '@/components/functional/Navbar/Navbar';

describe('Navbar', () => {
  const props = {
    title: 'a',
    description: 'b',
    onTitlebarClick: jest.fn(),
    onLeftButtonClick: jest.fn(),
  };

  test('should render without props', () => {
    const component = shallow(<Navbar />);
    expect(component).toMatchSnapshot();
  });

  test('should render with props', () => {
    const component = shallow(<Navbar {...props} />);
    expect(component).toMatchSnapshot();
  });

  test('should call onLeftButtonClicked when left button is clicked', () => {
    const component = shallow(<Navbar {...props} />);
    component.find('.Navbar__leftbutton').simulate('click');
    expect(props.onLeftButtonClick).toHaveBeenCalled();
  });

  test('should call onTitlebarClick when title bar is clicked', () => {
    const component = shallow(<Navbar {...props} />);
    component.find('.Navbar__titlebar').simulate('click');
    expect(props.onTitlebarClick).toHaveBeenCalled();
  });
});
