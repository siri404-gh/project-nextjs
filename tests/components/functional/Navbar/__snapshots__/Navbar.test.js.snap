// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render with props 1`] = `
<WithStyles(ForwardRef(AppBar))
  className="makeStyles-appBar-1"
  elevation={0}
>
  <WithStyles(ForwardRef(Toolbar))
    className="makeStyles-toolbar-7"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-hidden={true}
      className="Navbar__leftbutton makeStyles-leftBtn-2"
      onClick={[MockFunction]}
    >
      <MenuIcon
        color="secondary"
      />
    </WithStyles(ForwardRef(IconButton))>
    <div
      className="Navbar__titlebar makeStyles-titlebar-4"
      onClick={[MockFunction]}
    >
      <div
        className="makeStyles-titleLeft-3"
      >
        <Hidden
          smUp={true}
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-title-5"
            noWrap={true}
            variant="h6"
          >
            b
          </WithStyles(ForwardRef(Typography))>
        </Hidden>
        <Hidden
          xsDown={true}
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-title-5"
            noWrap={true}
            variant="h6"
          >
            a
          </WithStyles(ForwardRef(Typography))>
            |  
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-description-6"
            noWrap={true}
            variant="caption"
          >
            b
          </WithStyles(ForwardRef(Typography))>
        </Hidden>
      </div>
      <WithStyles(ForwardRef(IconButton))
        aria-controls="simple-menu"
        aria-hidden={true}
        onClick={[Function]}
      >
        <AccountCircleIcon
          className="makeStyles-small-8"
          color="secondary"
        />
        <WithStyles(ForwardRef(Menu))
          anchorEl={null}
          id="simple-menu"
          keepMounted={true}
          onClose={[Function]}
          open={false}
        >
          <Link
            href="/api/auth/login?redirectTo=/"
          >
            <WithStyles(ForwardRef(MenuItem))>
              Login
            </WithStyles(ForwardRef(MenuItem))>
          </Link>
        </WithStyles(ForwardRef(Menu))>
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;

exports[`Navbar should render without props 1`] = `
<WithStyles(ForwardRef(AppBar))
  className="makeStyles-appBar-1"
  elevation={0}
>
  <WithStyles(ForwardRef(Toolbar))
    className="makeStyles-toolbar-7"
  >
    <WithStyles(ForwardRef(IconButton))
      aria-hidden={true}
      className="Navbar__leftbutton makeStyles-leftBtn-2"
      onClick={[Function]}
    >
      <MenuIcon
        color="secondary"
      />
    </WithStyles(ForwardRef(IconButton))>
    <div
      className="Navbar__titlebar makeStyles-titlebar-4"
      onClick={[Function]}
    >
      <div
        className="makeStyles-titleLeft-3"
      >
        <Hidden
          smUp={true}
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-title-5"
            noWrap={true}
            variant="h6"
          />
        </Hidden>
        <Hidden
          xsDown={true}
        >
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-title-5"
            noWrap={true}
            variant="h6"
          />
            |  
          <WithStyles(ForwardRef(Typography))
            className="makeStyles-description-6"
            noWrap={true}
            variant="caption"
          />
        </Hidden>
      </div>
      <WithStyles(ForwardRef(IconButton))
        aria-controls="simple-menu"
        aria-hidden={true}
        onClick={[Function]}
      >
        <AccountCircleIcon
          className="makeStyles-small-8"
          color="secondary"
        />
        <WithStyles(ForwardRef(Menu))
          anchorEl={null}
          id="simple-menu"
          keepMounted={true}
          onClose={[Function]}
          open={false}
        >
          <Link
            href="/api/auth/login?redirectTo=/"
          >
            <WithStyles(ForwardRef(MenuItem))>
              Login
            </WithStyles(ForwardRef(MenuItem))>
          </Link>
        </WithStyles(ForwardRef(Menu))>
      </WithStyles(ForwardRef(IconButton))>
    </div>
  </WithStyles(ForwardRef(Toolbar))>
</WithStyles(ForwardRef(AppBar))>
`;
